transport-native-unix-common/src/main/java/io/netty/channel/unix/Unix.java

```diff
@@ -53,11 +53,9 @@
      * @param registerTask
      */
     @UnstableApi
-    public static void registerInternal(Runnable registerTask) {
-        if (registered.compareAndSet(false, true)) {
-            registerTask.run();
-            Socket.initialize();
-        }
+    public static synchronized void registerInternal(Runnable registerTask) {
+        registerTask.run();
```

Comment:
Can you confirm my understanding of the issue? Each netty-unix-common module (e.g. `netty_unix_errors_JNI_OnLoad`) which has two responsibilities:

1. RegisterNatives for all dynamic methods
2. Initialize static/global state in the current library

Since netty-unix-common is statically linked (2) needs to happen for each library that consumes it (e.g. epoll, kqueue, io_uring, ..), and (1) should only happen a single time per jvm (assuming the shading prefix is consistent for each library, and no conflicting netty versions on the class path). IIUC this change will make it such that both will always happen [1]. Once way to archives this would be each netty-unix-common module would have an additional overload to control conditional initialization.

```c
///// netty_unix_errors.h
jint netty_unix_errors_JNI_OnLoad(JNIEnv* env, const char* packagePrefix);
jint netty_unix_errors_JNI_OnLoad_Static(JNIEnv* env, const char* packagePrefix, bool registerNatives);

///// netty_unix_errors.c
jint netty_unix_errors_JNI_OnLoad(JNIEnv* env, const char* packagePrefix) {
  return netty_unix_errors_JNI_OnLoad_Static(env, packagePrefix, true);
}

jint netty_unix_errors_JNI_OnLoad_Static(JNIEnv* env, const char* packagePrefix, bool registerNatives) {
  // same code as in netty_unix_errors_JNI_OnLoad, but conditionally call netty_jni_util_register_natives
}

///// netty_unix.h
jint netty_unix_register(JNIEnv* env, const char* packagePrefix);
jint netty_unix_register(JNIEnv* env, const char* packagePrefix, bool registerNatives);

///// netty_unix.c
// similar changes as netty_unix_errors.c

// the unregister path would need to also track if the current library needs to unload natives
```

[1] EPOLL Native.java static init -> `Unix.registerInternal` -> EPOLL `registerUnix` -> `netty_epoll_native_registerUnix` -> `netty_unix_register` -> `netty_unix_errors_JNI_OnLoad` ... I

Reply:
yep thats it...

Reply:
I think with this change it should "just work".

Reply:
Won't the current state of this change will result in `RegisterNatives` on each invocation? I assumed this would result in `JNI_ERR` being returned and bailing out of initialization (I thought this was the original initiation of `registerInternal`)?

Reply:
Talked with @Scottmitch offline and all is good :)